File: web/src/lib/replay.ts
// Replay an event stream into per-step worlds.
//
// The trace JSON ships an `initialState` and a flat `events` list (see
// `julia/src/trace.jl`). One world is captured at every `step_marker`. Worlds
// use structural sharing: a `mut_set` rebuilds only the spine of the mutated
// path, so unchanged subtrees keep their object identity across steps. That
// makes `JSON.stringify(prev.X) === JSON.stringify(curr.X)` cheap for the
// cell-flash highlighting in `StateInspector`/`TapeInspector`.

import type {
  ArgValue,
  DebuggerState,
  InitialState,
  PathComponent,
  RValue,
  Root,
  TraceEvent,
} from "../types";

export interface World {
  ssa: ReadonlyMap<number, RValue>;
  args: readonly ArgValue[];
  tape: readonly RValue[];
  input: RValue;
}

export function buildWorlds(initial: InitialState, events: TraceEvent[]): World[] {
  let input: RValue = initial.input;
  let tape: readonly RValue[] = initial.tape;
  let args: readonly ArgValue[] = [];
  let ssa: ReadonlyMap<number, RValue> = new Map();
  const worlds: World[] = [];

  for (const e of events) {
    switch (e.t) {
      case "pass_start":
        args = e.args;
        ssa = new Map();
        break;
      case "ssa_define": {
        const next = new Map(ssa);
        next.set(e.pc, e.value);
        ssa = next;
        break;
      }
      case "mut_set": {
        const value = e.value;
        ({ input, args, tape, ssa } = applyAtRoot(
          { input, args, tape, ssa },
          e.root,
          e.path,
          () => value,
        ));
        break;
      }
      case "stack_push": {
        const item = e.value;
        ({ input, args, tape, ssa } = applyAtRoot(
          { input, args, tape, ssa },
          e.root,
          e.path,
          (stk) => {
            const items = (stk.items as RValue[]).concat([item]);
            return { kind: "stack", size: items.length, items };
          },
        ));
        break;
      }
      case "stack_pop":
        ({ input, args, tape, ssa } = applyAtRoot(
          { input, args, tape, ssa },
          e.root,
          e.path,
          (stk) => {
            const items = (stk.items as RValue[]).slice(0, -1);
            return { kind: "stack", size: items.length, items };
          },
        ));
        break;
      case "step_marker":
        worlds.push({ ssa, args, tape, input });
        break;
    }
  }
  return worlds;
}

/** Project a World into the legacy DebuggerState shape consumed by the UI. */
export function worldToDebuggerState(w: World): DebuggerState {
  const ssa = Array.from(w.ssa.entries())
    .sort(([a], [b]) => a - b)
    .map(([pc, value]) => ({ id: `%${pc}`, value }));
  return {
    ssa,
    args: w.args.map((a) => ({ id: a.id, role: a.role, value: a.value })),
    tape: w.tape.slice(),
    input: w.input,
  };
}

interface MutableWorld {
  input: RValue;
  args: readonly ArgValue[];
  tape: readonly RValue[];
  ssa: ReadonlyMap<number, RValue>;
}

function applyAtRoot(
  w: MutableWorld,
  root: Root,
  path: PathComponent[],
  f: (prev: RValue) => RValue,
): MutableWorld {
  switch (root.kind) {
    case "input":
      return { ...w, input: setAtPath(w.input, path, f) };
    case "arg": {
      const args = w.args.slice();
      const old = args[root.index];
      args[root.index] = { ...old, value: setAtPath(old.value, path, f) };
      return { ...w, args };
    }
    case "tape": {
      const tape = w.tape.slice();
      tape[root.index] = setAtPath(tape[root.index], path, f);
      return { ...w, tape };
    }
    case "ssa": {
      const ssa = new Map(w.ssa);
      const old = ssa.get(root.pc);
      if (old !== undefined) ssa.set(root.pc, setAtPath(old, path, f));
      return { ...w, ssa };
    }
  }
}

function setAtPath(value: RValue, path: PathComponent[], f: (prev: RValue) => RValue): RValue {
  if (path.length === 0) return f(value);
  const [head, ...rest] = path;
  if (typeof head === "string") {
    const k = value.kind;
    if (k === "codual") {
      const inner = value[head] as RValue;
      return { ...value, [head]: setAtPath(inner, rest, f) };
    }
    if (k === "tangent" || k === "namedtuple" || k === "struct") {
      const fields = (value.fields as { name: string; value: RValue }[]).map((fld) =>
        fld.name === head ? { name: fld.name, value: setAtPath(fld.value, rest, f) } : fld,
      );
      return { ...value, fields };
    }
    if (k === "ref") {
      const inner = value.value as RValue;
      return { ...value, value: setAtPath(inner, rest, f) };
    }
    throw new Error(`unexpected string path component '${head}' for kind '${k}'`);
  }
  const items = (value.items as RValue[]).slice();
  items[head] = setAtPath(items[head], rest, f);
  return { ...value, items };
}

File: MooncakeWalkthrough/src/replay.jl
# Replay an event stream back into per-step rendered worlds.
# Used only by the test suite — production tracing emits events; the browser
# (web/src/lib/replay.ts) replays them. This Julia replay mirrors that logic
# so tests can validate equivalence without a browser.

"""Replay an event stream into per-step rendered worlds (one world per
`step_marker`, in order)."""
function replay_worlds(initial_state::AbstractDict, events::AbstractVector)
    worlds = Dict{String,Any}[]
    input = deepcopy(initial_state["input"])
    tape = Any[deepcopy(s) for s in initial_state["tape"]]
    args = Any[]
    ssa = Dict{Int,Any}()
    for ev in events
        t = ev["t"]
        if t == "pass_start"
            args = Any[Dict{String,Any}("id" => a["id"], "role" => a["role"],
                                         "value" => deepcopy(a["value"]))
                       for a in ev["args"]]
            ssa = Dict{Int,Any}()
        elseif t == "ssa_define"
            ssa[ev["pc"]] = deepcopy(ev["value"])
        elseif t == "mut_set"
            _apply_root!(input, args, tape, ssa, ev["root"], ev["path"],
                         _ -> deepcopy(ev["value"]))
        elseif t == "stack_push"
            _apply_root!(input, args, tape, ssa, ev["root"], ev["path"],
                         stk -> Dict{String,Any}("kind" => "stack",
                                                  "size" => stk["size"] + 1,
                                                  "items" => vcat(stk["items"], [deepcopy(ev["value"])])))
        elseif t == "stack_pop"
            _apply_root!(input, args, tape, ssa, ev["root"], ev["path"],
                         stk -> Dict{String,Any}("kind" => "stack",
                                                  "size" => stk["size"] - 1,
                                                  "items" => stk["items"][1:end-1]))
        elseif t == "step_marker"
            world_ssa = Dict{String,Any}[Dict{String,Any}("id" => "%$pc",
                                                          "value" => deepcopy(ssa[pc]))
                                         for pc in sort(collect(keys(ssa)))]
            push!(worlds, Dict{String,Any}(
                "ssa" => world_ssa,
                "args" => Any[Dict{String,Any}("id" => a["id"], "role" => a["role"],
                                                "value" => deepcopy(a["value"]))
                              for a in args],
                "tape" => Any[deepcopy(s) for s in tape],
                "input" => deepcopy(input),
            ))
        end
    end
    return worlds
end

# Apply f to the value at root+path. The four roots have different shapes, so
# the dispatch lives here; `_set_at_path` handles the path-walk.
function _apply_root!(input, args, tape, ssa, root, path, f)
    rk = root["kind"]
    if rk == "input"
        new_val = _set_at_path(input, path, f)
        # `input` is a Dict — rebuild contents (cannot reassign in caller via this fn).
        empty!(input); merge!(input, new_val)
    elseif rk == "arg"
        i = root["index"] + 1
        new_inner = _set_at_path(args[i]["value"], path, f)
        args[i] = Dict{String,Any}("id" => args[i]["id"], "role" => args[i]["role"],
                                    "value" => new_inner)
    elseif rk == "tape"
        i = root["index"] + 1
        tape[i] = _set_at_path(tape[i], path, f)
    elseif rk == "ssa"
        pc = root["pc"]
        ssa[pc] = _set_at_path(ssa[pc], path, f)
    end
end

# Returns a NEW value with the leaf at `path` replaced by `f(prev_leaf)`.
function _set_at_path(value, path, f)
    if isempty(path)
        return f(value)
    end
    component = path[1]
    rest = path[2:end]
    if component isa AbstractString
        k = value["kind"]
        if k == "codual"
            new_inner = _set_at_path(value[component], rest, f)
            return merge(value, Dict{String,Any}(component => new_inner))
        elseif k == "tangent" || k == "namedtuple" || k == "struct"
            new_fields = Any[]
            for fld in value["fields"]
                if fld["name"] == component
                    push!(new_fields, Dict{String,Any}("name" => fld["name"],
                                                       "value" => _set_at_path(fld["value"], rest, f)))
                else
                    push!(new_fields, fld)
                end
            end
            return merge(value, Dict{String,Any}("fields" => new_fields))
        elseif k == "ref"
            new_inner = _set_at_path(value["value"], rest, f)
            return merge(value, Dict{String,Any}("value" => new_inner))
        else
            error("unexpected string path component '$component' for kind '$k'")
        end
    else  # number, 0-based
        i = component + 1
        new_items = copy(value["items"])
        new_items[i] = _set_at_path(new_items[i], rest, f)
        return merge(value, Dict{String,Any}("items" => new_items))
    end
end
